ruby/lib/google/protobuf/repeated_field.rb

```diff
@@ -29,11 +29,127 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # add syntatic sugar on top of the core library
+require 'forwardable'
 module Google
   module Protobuf
     class RepeatedField
+      extend Forwardable
 
-      alias_method :size, :length
+      # methods defined in C or Java:
+      #   +
+      #   [], at
+      #   []=
+      #   concat
+      #   clear
+      #   dup, clone
+      #   each
+      #   hash
+      #   push, <<
+      #   replace
+      #   length, size
+      #   ==
+      #   to_ary, to_a
+      #   also all enumerable
+      #
+      # NOTE:  using delegators rather than method_missing to make the
+      #        relationship explicit instead of implicit
+      # NOTE2: use to_a instea
+      def_delegators :to_ary,
+        :&, :*, :-, :'<=>',
+        :assoc, :bsearch, :combination, :compact, :count, :cycle,
+        :drop, :drop_while, :eql?, :fetch, :find_index, :flatten,
+        :include?, :index, :inspect, :join,
+        :pack, :permutation, :product, :pretty_print, :pretty_print_cycle,
+        :rassoc, :repeated_combination, :repeated_permutation, :reverse, :rindex, :rotate,
+        :sample, :shuffle, :shelljoin, :slice, :to_s, :transpose, :uniq, :|
+
+
+      def first(n=nil)
+        n ? self[0..n] : self[0]
+      end
+
+
+      def last(n=nil)
+        n ? self[(self.size-n-1)..-1] : self[-1]
+      end
+
+
+      def pop(n=nil)
+        if n
+          results = []
+          n.times{ results << pop_internal }
+          return results
+        else
+          return pop_internal
+        end
+      end
+
+
+      def empty?
+        self.size == 0
+      end
+
+      # array aliases into enumerable
+      alias_method :each_index, :each_with_index
+      alias_method :slice, :[]
+      alias_method :values_at, :select
+      alias_method :map, :collect
+
+
+      %w(delete delete_at delete_if shift slice! unshift).each do |method_name|
+        define_method(method_name) do |*args, &block|
+          arr = self.to_a
+          result = arr.send(method_name, *args)
+          self.replace(arr)
```

Comment:
so I implemented an unshift; I figured that may be one of the heaviest calls.  Here are the stats:
```
# ruby-custom unshift
  - mem before: 36.1
  - first 1,000: 0.0025720237754285336
  - first 1,000 mem: 36.7
  - next 10,000 mem: 36.5
  - next 10,000: 0.037256487645208836

# c-custom unshift
  - mem before: 36
  - first 1,000 mem: 36.2
  - first 1,000: 0.000838711392134428
  - next 10,000 mem: 36.2
  - next 10,000: 0.005491869989782572

# Array unshift
  - mem before: 31.2
  - first 1,000 mem: 31.6
  - first 1,000: 0.00038467906415462494
  - next 10,000 mem: 36.2
  - next 10,000: 0.007755566854029894
```

The problem was that `RepeatedField#unshift` segfaulted after 32 strings were added.  I suspect this is a 256 bits issues, as the element_size for a string is 8, and that I wasn't expanding memory correctly.  That is why there is wonky benchmarking code (see below).  But the above results are probably indicative of the performance degradation on some of these calls, except for memory, which will explode for very large repeated_field sizes.  

one way to do this may be to keep the ruby code, and add a TODO to optimize a few of these really heavy ones (like unshift)

-----------------------
### code
```c
VALUE RepeatedField_unshift(VALUE _self, VALUE val) {
  RepeatedField* self = ruby_to_RepeatedField(_self);
  upb_fieldtype_t field_type = self->field_type;
  int element_size = native_slot_size(field_type);
  if (self->size > 0){
    void* mem = RepeatedField_index_native(_self, self->size-1);
    RepeatedField_push_native(_self, mem);
  }
  size_t off = element_size * (self->size-3);
  int i = self->size-3;
  for (; i >= 0; i--, off -= element_size) {
    void* from_mem = (uint8_t *)self->elements + off;
    void* to_mem = (uint8_t *)self->elements + off + element_size;
    native_slot_dup(field_type, to_mem, from_mem);
  }
  RepeatedField_index_set(_self, INT2NUM(0), val);
  return _self;
}
```


### benchmarking code
```ruby
times = []
m = TestMessage.new
100.times do
  m.repeated_string.clear
  times << Benchmark.realtime do 
    10.times{|i| m.repeated_string.unshift(i.to_s) }
  end
end
times.inject{|sum,x| sum + x }

times = []
m = TestMessage.new
1_000.times do
  m.repeated_string.clear
  times << Benchmark.realtime do 
    10.times{|i| m.repeated_string.unshift(i.to_s) }
  end
end
times.inject{|sum,x| sum + x }
```